MM-42146-rhs-set-tasks-due-date

gbochora opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository-specific documentation at https://developers.mattermost.com/contribute/getting-started/

REMEMBER TO:
- Run `make i18n-extract` and commit changes to synchronize any new or removed messages
- Run `make check-style` to check for style errors (required for all pull requests)
- Run `make test` to ensure unit tests passed
-->

#### Summary
This PR adds `Add due date` functionality to the RHS checklist item. Customers can specify any date/time in the input or choose from the suggested list.
If a due date is set for the checklist item, the date selector will show it at the end of the suggested list.


https://user-images.githubusercontent.com/4368372/159030768-5ca449c9-7308-4611-87b4-337b63164f90.mp4


Missing e2e tests, will add them after https://mattermost.atlassian.net/browse/MM-42080 is done.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42146
https://mattermost.atlassian.net/browse/MM-42152

#### Checklist
<!-- Check off items as they are completed. ~~Strike through~~ items if they don't apply -->
- ~~[ ] Telemetry updated~~
- ~~[ ] Gated by experimental feature flag~~
- [ ] Unit tests updated

calebroseland reviewed (changes requested):
Aside from a small issue or two, this is really good! Nice job 😃

1. Changing the due date (or clearing it) doesn't reflect selected state until you mouse out and back in (to cause a re-render?)
![CleanShot 2022-03-24 at 23 22 28](https://user-images.githubusercontent.com/11724372/160056463-a9b874d3-2ef0-4329-8f9f-a193c201528e.gif)
![CleanShot 2022-03-24 at 23 41 44](https://user-images.githubusercontent.com/11724372/160056342-60b7bbf2-614e-4fa4-ac1a-edf1c85b0838.gif)

2. More of a UX observation, maybe instead of just`Today`, the closer to "now" could perhaps be more granular options—like `in 4 hours`, `this afternoon`, `tonight`, `tomorrow morning`, etc. (1/5; and more for discussion and potential follow up ticket.)


On code reuse, I like that some `datetime_input` internals are shared, but I wonder how tied together they should be—if that might cause issues down the road. Should they be the same component with different UI modes? (Traditional input, dropdown-style, etc.) Nothing needs changing now of course, just some thoughts we could work through perhaps.

I really like the following designs with the integrated calendar/time input (cc @abhijit-singh 🙌) and even before seeing want to eventually add a manual date/time selector to `datetime_input` as well. There's a somewhat interesting [react-select example here](https://react-select.com/advanced#experimental) on this.
<img width="456" alt="CleanShot 2022-03-24 at 23 14 17@2x" src="https://user-images.githubusercontent.com/11724372/160057646-abd9780d-fc09-4b64-ae9d-2f26a7291558.png">

abhijit-singh reviewed (commented):
Thanks, @gbochora! Great work!

Aside from the update issue @calebroseland mentioned, after selecting a date, the task actions container seems to persist until hovered on and off again. Not sure if this issue was there earlier also though. 
![2022-03-25 17 38 00](https://user-images.githubusercontent.com/14369607/160118333-94920a1a-ff83-4db6-98d3-502fb1adf52a.gif)

iomodo commented on webapp/src/components/checklist_item/duedate.tsx:
```diff
@@ -0,0 +1,133 @@
+        }
+    );
+
+    dateTime = dateTime.plus({days: 7});
```
`dateTime` will be plus 8 here not 7 right?

gbochora replied:
nice catch!

iomodo commented on webapp/src/components/checklist_item/duedate.tsx:
```diff
@@ -0,0 +1,133 @@
+	font-size: 22px;
+`;
+
+export default DueDate;
```
Missing enter

iomodo reviewed (approved):
Looks great! Minor suggestions before merging.

gbochora commented:
@abhijit-singh @calebroseland @iomodo Thanks for the feedback.
Changelog:

- Fixed task actions menu persist issue. Added hiding menu to `onSelectedChange` callbacks. 
- Fixed options list update issue. `useMemo` prevented rerendering the options list.
- Moved code from `check_list_item` to the new `DueDate` component to make it easier to merge with the new checklist branch.

calebroseland reviewed (changes requested):
Sorry I forgot to check previously, but if you enter a duration (e.g. `89 days`), `chrono-node` won't pick it up and it'll fallback to `parse-duration` (by design, the same happens in `datetime_input`).

When this happens, `OptionLabel` in `datetime_selector` does `DateTime.now().plus //...`, but over in `checklist_item`, `onDueDateChange` still receives the raw duration, resulting in due dates in the 1970s and such. 

<img width="300" alt="image" src="https://user-images.githubusercontent.com/11724372/160531346-15bb017b-bba6-48f6-867e-99d5c8cef2bd.gif">

abhijit-singh reviewed (approved):
Looks good to me! Thanks @gbochora!

gbochora commented on webapp/src/components/checklist_item.tsx:
```diff
@@ -413,6 +417,22 @@ export const ChecklistItemDetails = (props: ChecklistItemDetailsProps): React.Re
+        }
+        let timestamp = 0;
+        if (value?.value) {
+            timestamp = value?.value.toMillis();
```
I think an issue here is that `defaultMakeOptions` returns an option which's type is `DateTime`, but the value is duration. It seems correct that in `Mode.DateTime` value should always be a specific date, not duration.
I've changed `defaultMakeOptions`, now it creates `DateTime` value from duration and returns. 
Could this change break something I'm missing? If yes, then I'll do the way you suggested.

calebroseland replied:
Working great and a better place to fix this!

calebroseland replied:
I tested the other locations and it doesn't break anything. Since your change is semantic according to `Mode`, it seems to work nicely. 💯
